std/algebra/twistededwards/bandersnatch/point.go

```diff
@@ -98,41 +107,87 @@ func (p *Point) Double(api frontend.API, p1 *Point, curve EdCurve) *Point {
 	return p
 }
 
+// phi endomorphism √-2 ∈ 𝒪₋₈
+// (x,y) → λ × (x,y) s.t. λ² = -2 mod Order
+func (p *Point) phi(api frontend.API, p1 *Point, curve EdCurve) *Point {
+
+	xy := api.Mul(p1.X, p1.Y)
+	yy := api.Mul(p1.Y, p1.Y)
+	f := api.Sub(1, yy)
+	f = api.Mul(f, curve.endo1)
+	g := api.Add(yy, curve.endo0)
+	g = api.Mul(g, curve.endo0)
+	h := api.Sub(yy, curve.endo0)
+
+	p.X = api.DivUnchecked(f, xy)
+	p.Y = api.DivUnchecked(g, h)
+
+	return p
+}
+
+var scalarDecompositionHint = hint.NewStaticHint(func(curve ecc.ID, inputs []*big.Int, res []*big.Int) error {
+	// TODO: handle properly negative scalars
+	var lambda, order big.Int
+	var glvBasis ecc.Lattice
+	lambda.SetString("8913659658109529928382530854484400854125314752504019737736543920008458395397", 10)
+	order.SetString("13108968793781547619861935127046491459309155893440570251786403306729687672801", 10)
+	ecc.PrecomputeLattice(&order, &lambda, &glvBasis)
+	sp := ecc.SplitScalar(inputs[0], &glvBasis)
+	res[0].Neg(&(sp[0])) // Set
+	res[1].Set(&(sp[1]))
+
+	// figure out how many times we have overflowed
+	// res[2].Mul(res[1], &lambda).Add(res[2], res[0])
+	res[2].Mul(res[1], &lambda).Sub(res[2], res[0])
+	res[2].Sub(res[2], inputs[0])
+	res[2].Div(res[2], &order)
+
+	return nil
+}, 1, 3)
+
+func init() {
+	hint.Register(scalarDecompositionHint)
```

Comment:
hint func should be public, unless I missed something in the new pattern by @ivokub . 
Else, how would a prover (which may not import this package) provide the hint function to the backend calls?

Reply:
This is on me and sw implementation has the same problem. If the package isn't imported (both the CS and witness are deserialized), then the hint won't be provided.

In addition to exporting, I think doing something like #264 would also help with usability a bit?

Reply:
I followed the pattern in SW. Let's keep this in a separate ticket.